### DISCUSSION TITLE
export alpine data using imported things and use it in other module

### DISCUSSION THREAD

**Discussion by Shashank-Bhatt1** [alpinejs/alpine#3883] [1. Help] [answered]
This might not be probably the alpine js question but as I am first time using this with bundler way, i am kind of confused to decide what is the proper way to define data if data is using imported variables. 

I genuinely appreciate the help over this platform and just loved to use this amazing library.

I am doing as suggested in the doc [here](https://alpinejs.dev/globals/alpine-data#registering-from-a-bundle) and exporting data as a function returning object.

However, in my case alpine js data itself is using imported things from other modules so let;s say in my `main.js` i imported Alpine and the html.

```
import Alpine from 'alpinejs';
import html from './main.html';
import initStyles from '../../initStyles';
import initHTML from '../../initHTML';
```
and then exporting the function like this 

```
export default () => ({
    init() {
        initHTML('main-component',html);
        initStyles(this.$el.shadowRoot);
    },
    initText: 'Hello World!!'  
}))
```

and then in my index.js i am importing the data which i exported  from main.

```
import main from main;
Alpine.data('main', main)

```

and then when i try to use main as x-data in html i am getting error that `html` is not defined here as it was only scoped in `main,js` where it was imported.
`<div x-data="main"></div>`

I saw that it just replaces function `expression ()=>{return {..}}` in place of main in html but as it is scoped there now in html, html variable is not defined.

I am using now Alpine.data in the same module (main.js) instead of exporting function to kind of get rid of this so instead of exporting anything, i use.

```
Alpine.data('main', () => ({
    init() {
        initHTML('main-component',html);
        initStyles(this.$el.shadowRoot);
    },
    initText: 'Hello World!!'  
}))
```

but i believe this is not the proper way to define as being module it is not exporting anything in that case. I kind of want to use how it is suggested in [documentation](https://alpinejs.dev/globals/alpine-data#registering-from-a-bundle) to define data after it is imported in index.js (dropdown example). or is it that it doesn;t  need to export anything and i can also use without export? how?

Can i have these imported things (`html` variable here) defined when i use that `main` data in html?

**Comment by ekwoka** [comment]
Can you give more context? What bundler? What does the code actually look like for real? Etc.

100% the first one works.

Content is scoped. Your data function can access variables defined in it's module scope.

Alpine doesn't change anything about it.

**Reply by shashank-brightness** [reply]
@ekwoka here is how everything is there. I am using now Alpine.data in same main in place of exporting. I believe it is something that i am missing while exporting. May be html and all other things which are imported?
Content from main.js

```
import Alpine from 'alpinejs';
import html from './main.html';
import initStyles from '../../initStyles';
import initHTML from '../../initHTML';

export default () => ({
    init() {
        initHTML('main-component',html);
        initStyles(this.$el.shadowRoot);
    },
    initText: 'Hello World!!'  
})
```

here is the content from `index.js` of my own package where main is imported and also whole component is returned to create dom to append in html.

```
import main from './components/main/main';
Alpine.data('main', main)
Alpine.start();

function getMainComponent() {
  return `
    <div  x-data="localization">
      <main-component datatemplateurl="main/main.html" x-data="${main}" styles="global" shadow="true"></main-component>
    </div>
  `;
} 
export {getMainComponent};
```

and then finally somewhere in other package it is required

```
import {getMainComponent} from 'my-package-developed-with-alpine';

if(!document.querySelector('x-component')) {
    document.body.insertAdjacentHTML('beforeend',getMainComponent());
}
```

and i am gettting whole export function inside the dom where it says html is not defined. As we see it changes function to webpack modules defination like __webpack__module[something] but then html is still not defined there/
![Screenshot 2023-11-27 131628](https://github.com/alpinejs/alpine/assets/141903260/155ba903-c2ff-4e75-a8a1-320a1e4937c2)

**Reply by ekwoka** [reply, marked as answer]
so, the issue here is that you literally put the FUNCTION into the template.

When you just need to put the string `'main'`

Additionally, if it's not too late, PLEASE PLEASE PLEASE use Vite.

You should also probably not be calling Alpine start in the package you are exporting from. Almost certainly not.

You should simply be exporting a function that accepts `Alpine` as an arg, so it can be passed to `Alpine.plugin` at the consumer side.

similarly, you can register a directive to handle `x-component` with alpine.

**Reply by shashank-brightness** [reply]
@ekwoka  Thanks for the prompt response. Ah so the thing is that if i use the string here like `x-data="main"`, it works. Cool.

I still have two questions if you can help.

1. I am already using Alpine js [plugin](https://github.com/markmead/alpinejs-component) called `alpine-component` in my own package(dependency of that) which already does something you mentioned. i.e. take Alpine as a parameter and returning function so i used like that.

`import component from 'alpinejs-component';
Alpine.plugin(component)`

which was called `x-component` 

But i needed my own names(e.g. main-component) and my own connectedCallback and i can't change package, so I went through that package's source code and extended that and built my own components on top of it, If i need to return entire html with let;s say custom tags mentioned in `getMainComponent`, how would I do that? 
My core logic has already been done now in my package and my goal is to get the entire html built with functionality (not just the functionality) when someone requires my package at consumer side. 
What I am doing is now first adding main-component mounted in dom and then subsequently it will nest other components in it;s shadow dom and so on. All custom components will have it's own shadow dom. I did not know how i get html as module(string of it) earlier so i configured webpack, 

Regarding vite, i will try to use it in next as this package needs to be configured somewhere where webpack is going to be used, as it will take some time to configure and get my feet on vite just when i just completed that in webpack. :) I just read it;s doc and it says it has been built with keeping speed as the main thing so would try it later.

2.  If i register data via Alpine.data method, how would i get my modified data at some point in other modules at that time? Let's say in my other modules i need to get main component's $refs magic properties, (because at that point alpine has initialized data in the tree so i need that `x-ref` of main data which can point to some element in the tree where the `main` data is spread across.) Alpine.data defines data then how to get  the live `main` data (not just the exported function) in other modules? 
I need this as my all components are under custom shadow root so i can't use querySelector/All methods. 

Note: (I must have shadow root in order to get style encapsulation which i didn;t know earlier but thanks to [creator](https://github.com/markmead) of x-component)

**Reply by ekwoka** [reply]
> as it will take some time to configure and get my feet on vite just when i just completed that in webpack

I can bet it will almost take no configuration except turning on lib mode.

I'm not quite sure I understand. Why would you need to get data from this component not within an alpine context tree?

While not really recommended, you can do something like

```js
const mainData = Alpine.reactive({
  count: 0
})

Alpine.data('main', () => mainData)
```

which would give you a single shared object, but really, at that point, why not use a store?

It's a bit hard to tell what you actually want to do, but almost definitely the idea of needing x-ref to refer to a specific alpine component somewhere is a fundamental design issue.

I'm also not totally sure what you mean by "add my own components"? It looks like that package lets you.

**Reply by shashank-brightness** [reply]
I have already taken one store where i need to use global functions. Agreed on that but there might be the cases where there is an element inside the dom and i need to be able to refer that element somewhere else to initalize plugin on that and even resize on events like browser resize or modal popup open event. I am even using bootstrap plugins in them.

For example, i am working on canvas library and i need to resize canvas based on parent element's width as that;s the one which i will use as base to resize because the main div where my x-data is there can be hidden/shown on conditions. That;s the thing why i need to refer that element in other module where everything is inside it;s own shadowRoot,.

My own components means the [web-components](https://developer.mozilla.org/en-US/docs/Web/API/Web_components) same as `x-component` does but with custom names.

kind of like this (initHTML.js is this)

```
import Alpine from 'alpinejs';

export default function(componentName, componentHTML) {
  class xCustomComponent extends customElements.get('x-component') {
    constructor() {
      super();
    }
    connectedCallback() {
      
      const shadow = this.getAttribute('shadow');
      //const shadow = false;
      
      const shadowRoot = shadow ? this.attachShadow({ mode: 'open' }) : null;
      

      const newComponent = new DOMParser().parseFromString(
        componentHTML,
        'text/html'
      ).body.firstChild;
      
      const alpineInitPointEl = shadow ? shadowRoot : this; 
  
      alpineInitPointEl.appendChild(newComponent);
      Alpine.initTree(alpineInitPointEl);
      
    }
  }   
  
  if (window.customElements.get(componentName)) {
    return
  }
     
  customElements.define(componentName,xCustomComponent);
      
}
```
   
I am not sure where is the documentation for initTree but i believe it initializes a tree on dom so i copied that from the plugin and extended it;s class.

and then on each individual component's js i would use like this.

```
export default () => ({
    init() {
        initHTML('main-component',html);
        initStyles(this.$el.shadowRoot);
    },
    initText: 'Hello World!!'  
})
```
so finally i can write in html `<main-component x-data="main" shadow="true"></main-component>` and reuse this anywhere.

**Reply by ekwoka** [reply]
> i need to resize canvas based on parent element's width as that;s the one which i will use as base to resize

Why would this be done from somewhere other than the canvas itself?

**Reply by shashank-brightness** [reply]
canvas is manipulated by some other [library](http://fabricjs.com/) and i need parent container (which resides in parent component) width  and even modal's main div element (which is in modal component) to handle that specific case (modal open and browser resize). 
I mean this is just the example but i might even need data's other properties as well in another component/s.

**Reply by shashank-brightness** [reply]
I will figure it out to create and share you the whole thing from my local to some online platform like stackblitz.

**Reply by ekwoka** [reply]
Mostly store is the way, and just you can make an object that you attach the elements to, or functions you can have the component/element pass itself to.
